server/src/main/java/org/elasticsearch/index/translog/TranslogStreamOutput.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.translog;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefIterator;
+import org.elasticsearch.common.io.stream.RecyclerBytesStreamOutput;
+import org.elasticsearch.common.recycler.Recycler;
+
+import java.io.IOException;
+
+/**
+ * This class provided specialized serialization methods for translog operations. The goal is to allow direct
+ * access to the current recycler bytes stream output page and write the entire operation with a single bounds
+ * check.
+ */
+public class TranslogStreamOutput extends RecyclerBytesStreamOutput {
+
+    public static final int FIXED_INDEX_HEADER_SIZE = 39;
+    public static final int FIXED_DELETE_HEADER_SIZE = 30;
+    public static final int FIXED_NO_OP_HEADER_SIZE = 21;
+
+    public TranslogStreamOutput(Recycler<BytesRef> recycler) {
+        super(recycler);
+    }
+
+    public void writeSerializedOperation(Translog.Serialized serialized) throws IOException {
+        final int preWritePageOffset = this.currentPageOffset;
+        final BytesRefIterator refIterator = serialized.iterator();
+        if (serialized.length() <= (pageSize - preWritePageOffset)) {
+            BytesRef currentPage = currentBytesRef;
+            int offset = currentPage.offset + preWritePageOffset;
+            byte[] pageBytes = currentPage.bytes;
+            BytesRef ref;
+            while ((ref = refIterator.next()) != null) {
+                System.arraycopy(ref.bytes, ref.offset, pageBytes, offset, ref.length);
+                offset += ref.length;
+            }
+            VH_BE_INT.set(pageBytes, offset, serialized.checksum());
+            this.currentPageOffset = preWritePageOffset + serialized.length();
+        } else {
+            BytesRef next;
+            while ((next = refIterator.next()) != null) {
+                writeBytes(next.bytes, next.offset, next.length);
+            }
+            writeInt(serialized.checksum());
+        }
+    }
+
+    public void writeIndexHeader(Translog.Index indexOperation) throws IOException {
+        final int currentPageOffset = this.currentPageOffset;
+        if (FIXED_INDEX_HEADER_SIZE <= (pageSize - currentPageOffset)) {
```

Comment:
Instead of subclassing, I wonder if we could have a "fastwrite(Consumer<FastWrite> writer, int size)" on `RecyclerBytesStreamOutput` that does the bounds checking once and exposes one of two objects (avoid mega morphic) where one will do no bounds checking and the other just does the regular write. That way we avoid sub classing `RecyclerBytesStreamOutput` (and all the protected variables). I do see that it generates an object, but perhaps it does not matter (escape analysis or just not important)?